Extending `replace_picture()` to scale images:

m-entrup opened the pull request:
Large images are always scaled.
The aspect ratio is kept.
The new kwarg `do_not_scale_up` can disable upscaling.
Centering is always active.

m-entrup committed:
Extending `replace_picture()` to scale images:

Large images are always scaled.
The aspect ratio is kept.
The new kwarg `do_not_scale_up`can disable upscaling.
Centering is always active.

m-entrup committed:
Changes according to the review of PR #6:

Raising an error if `shapes_to_replace` is empty.
Wrapping `open()`into a `with` statement.

m-entrup commented:
I was not able to include these changes into PR #6. Maybe it is best to close PR #6 and use this one instead. The change requests from PR #6 are already included.

michaelosthege commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,47 @@ def replace_text(self, label: str, new_text: str) -> None:
             label (str): label of the placeholder (without curly braces)
             filename (path-like): path to an image file
+            do_not_scale_up (bool): deactivates that the image is enlarged (default: False)
         """
```
A default kwarg `scale_up=True` would be easier to understand

m-entrup replied:
I have replaced `do_not_scale_up` by ``scale_up` and changed how it's used.
Additionally, I added some more information to the doc string of `replace_image()`.

timhoffm commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,47 @@ def replace_text(self, label: str, new_text: str) -> None:
+            raise ValueError(f"The label '{label}' can't be found in the template.")
+        if isinstance(filename, str):
+            filename = pathlib.Path(filename)
+        if not filename.is_file():
```
`open`raises an almost identical exception. `FileNotFoundError: [Errno 2] No such file or directory: 'abc'`. So, we do not need to check this explicitly.

m-entrup replied:
The newest commit removes the manual raise of an exception and lets `open()` handle non existing files.

timhoffm commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,47 @@ def replace_text(self, label: str, new_text: str) -> None:
+        shapes_to_replace = self._find_shapes(label)
+        if not shapes_to_replace:
+            raise ValueError(f"The label '{label}' can't be found in the template.")
+        if isinstance(filename, str):
```
Since the only usage of filename is `open(filename)` (see also comment below), we do not need an explicit conversion. Open can handle any path-like object.

m-entrup replied:
I removed the explicit conversion of `str` to `Path` and let `open()` handle it implicitly.

lysnikolaou commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,47 @@ def replace_text(self, label: str, new_text: str) -> None:
+            do_not_scale_up (bool): deactivates that the image is enlarged (default: False)
         """
-        pass
+        shapes_to_replace = self._find_shapes(label)
```
The method signature of `_find_shapes` was changed in #7. Maybe want to change this call to reflect that change?

m-entrup replied:
I have updated the usage of `_find_shapes()`. The same two lines of code are used that can be already be found at `replace_text()`.

timhoffm commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,47 @@ def replace_text(self, label: str, new_text: str) -> None:
             shape.text = new_text
 
-    def replace_picture(self, label: str, filename: _Pathlike) -> None:
+    def replace_picture(self, label: str, filename: _Pathlike, *, do_not_scale_up: bool = False) -> None:
```
What is the motivation / usecase szenario for `do_not_scale_up`?

I would want the  following behaviors in the future (might be in a future PR).
~~~
scale:
    All image scaling preseves the aspect-ratio of the original image.
    - 'fit-width': Scale the image to the width of the placeholder.
    - 'fit-height': Scale the image to the height of the placeholder.
    - 'fit': Scale the image to fit inside the placeholder.
~~~

Note, that `do_not_scale_up` and `scale` are sort of orthogonal.

Naming: One might want to use `scale_policy` instead of `do_not_scale_up`; with values "shrink-only", "enlarge-only", `None`.
~~~
scale_policy

m-entrup replied:
With my implementation the image does always fit into the size of the replaced shape. By calculating the aspect ratio I am able to defect if `height` or `width` is the limiting dimension. This is the same as `scale='fit'`.
I don't like the idea of adding the options 'fit-width' and 'fit-height'. Allowing the image to get larger than the placeholder will result in overlapping with other shapes.
The `scale_up=False` is useful if an image is only a little to small. Scaling it up by a view percent will reduce the image quality. By preventing up scaling we will get the best image quality.

timhoffm replied:
> I don't like the idea of adding the options 'fit-width' and 'fit-height'. Allowing the image to get larger than the placeholder will result in overlapping with other shapes.

Not saying that they should be default. But we shouldn't assume too much on the user's intention. I can imagine situations in which you'd really want to force the width/height - no matter the cost.

I agree that preventing upscaling might be desirable. Similarly, somebody might want to prevent downscaling since it reduces the information in the image.

m-entrup committed:
Change usage of _find_shapes() at replace_image():

Splitting up the label first using `_parse_label()`.
Using the results as inputs for `_find_shapes()`.

m-entrup committed:
Changing the kwarg `do_not_scale_up`:

`do_not_scale_up=False`is replaced by `scale_up=True`.
Code using the kwarg is changed to use the negation.
The doc string is extended to explain the `replace_image()` better.

m-entrup committed:
Using the functionality build into `open()`:

`open()` can handle path-like objects. We do not need to convert `str` to `Path`.
If a file does not exist òpen()` raises a meaningful exception.
It is not necessary to raise a `FileNotFoundError` manually.

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,49 @@ def replace_text(self, label: str, new_text: str) -> None:
 
-    def replace_picture(self, label: str, filename: _Pathlike) -> None:
+    def replace_picture(self, label: str, filename: _Pathlike, *, scale_up: bool = True) -> None:
         """Replaces rectangle placeholders on one or many slides.
```
Minor rewording please: ... on one or many slides with an image.

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,49 @@ def replace_text(self, label: str, new_text: str) -> None:
         """Replaces rectangle placeholders on one or many slides.
 
+        The aspect ratio of the image is not changed.
+        To large images are always resized.
```
To -> Too

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,49 @@ def replace_text(self, label: str, new_text: str) -> None:
+                    left=old_shape.left,
+                    top=old_shape.top,
+                )
+                # Scaling the image if `scale_up == True`:
```
According to the DocString, images too large for the old_shape are resized to fit into old_shape. This feature is still missing!

cmahr replied:
I really think that (at least) scaling the image should be delegated to a private method to improve readability.

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,49 @@ def replace_text(self, label: str, new_text: str) -> None:
+                # Centering the image at the extent of the placeholder:
+                img_shape.top += int((old_shape.height - img_shape.height) / 2)
+                img_shape.left += int((old_shape.width - img_shape.width) / 2)
+                del slide_shapes[slide_shapes.index(old_shape)]
```
For me running on Python 3.6.9 (installed via pyenv, test suite is OK), the deletion is not working:

```
Traceback (most recent call last):
  File "/home/carsten/.PyCharmCE2019.2/config/scratches/scratch.py", line 7, in <module>
    tpl.replace_picture('*:logo', '/home/carsten/Downloads/beer.jpg')
  File "/home/carsten/Documents/Python/pptx-blueprint/pptx_blueprint/__init__.py", line 75, in replace_picture
    del slide_shapes[slide_shapes.index(old_shape)]
TypeError: 'SlideShapes' object doesn't support item deletion
```

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -33,14 +33,49 @@ def replace_text(self, label: str, new_text: str) -> None:
+                if img_shape.height <= old_shape.height and img_shape.width <= old_shape.width and scale_up:
+                    old_aspect_ratio = old_shape.width / old_shape.height
+                    new_aspect_ratio = img_shape.width / img_shape.height
+                    if old_aspect_ratio >= new_aspect_ratio:
```
Using an image in portrait mode, the width of the image is set to the width of old_shape, which results in an overflow in y direction. This logic seems broken?

cmahr replied:
Vice versa, an image in landscape mode overflows in x direction. Replacing `>=` with `<` hence fixes this issue.
